src/charm.py

```diff
@@ -63,21 +63,21 @@ def __init__(self, *args):
         self._lightkube_field_manager = "lightkube"
         self._k8s_resource_handler = None
 
-        self.framework.observe(self.on.config_changed, self.service_patcher._patch)
-        self.framework.observe(self.on.remove, self._on_remove)
+        # setup events to be handled by main event handler
+        self.framework.observe(self.on.upgrade_charm, self._on_event)
+        self.framework.observe(self.on.config_changed, self._on_event)
+
+        for rel in self.model.relations.keys():
+            self.framework.observe(self.on[rel].relation_changed, self._on_event)
 
-        for event in [
-            self.on.install,
-            self.on.leader_elected,
-            self.on.upgrade_charm,
-            self.on["kubeflow-profiles"].relation_changed,
-            self.on.config_changed,
-        ]:
-            self.framework.observe(event, self.main)
+        # setup events to be handled by specific event handlers
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(self.on.remove, self._on_remove)
+        self.framework.observe(self.on.config_changed, self.service_patcher._patch)
```

Comment:
At line 68 above `config_changed` is setup with `_on_event`. We need to review this.

Reply:
I think an event can have multiple observers, how can we check on that?

Reply:
Yes, it looks like we can have multiple observers.

Reply:
Are we happy with the order of those observers? I.e. order they will be called when event is triggered.

Reply:
patch first then main would make sense. will adjust

Reply:
I suspect `_patch` is added to observer's list automatically when you instantiate `KubernetesServicePatch`:
https://github.com/canonical/kubeflow-profiles-operator/blob/main/lib/charms/observability_libs/v1/kubernetes_service_patch.py#L211

I think, we do not need to explicitly add.